File: ts/128/113.ts
import { TreeNode } from '../class'

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  const ans: number[][] = []
  if (!root) return ans

  const dfs = (node: TreeNode, target: number, paths: string = ''): void => {
    paths ? paths += `,${node.val}` : paths += node.val
    if (!node.left && !node.right && target === node.val) {
      ans.push(paths.split(',').map(v => +v))
    }
    if (node.left) dfs(node.left, target - node.val, paths)
    if (node.right) dfs(node.right, target - node.val, paths)
  }

  dfs(root, targetSum)
  return ans
};

File: py3/128/113.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans, path = [], []
        if not root: return ans

        def dfs(node, target):
            if not node: return
            target -= node.val
            path.append(node.val)
            if not node.left and not node.right and target == 0:
                ans.append(path[:])

            dfs(node.left, target)
            dfs(node.right, target)
            path.pop()

        dfs(root, targetSum)
        return ans
